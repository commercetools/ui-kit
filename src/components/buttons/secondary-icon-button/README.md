# Buttons: SecondaryIconButton

## Usage

```js
import { SecondaryIconButton } from '@commercetools-frontend/ui-kit';
```

#### Description

Secondary Icon Buttons are "icon-only" buttons and a restricted version of the
`<IconButton>`. They trigger an action when clicked (`onClick` prop). You must
also pass a label for accessibility reasons.

#### Usage

```js
<SecondaryIconButton
  icon={<ArrowRightIcon />}
  label="Next"
  onClick={() => alert('Button clicked')}
/>
```

#### Properties

| Props        | Type     | Required | Values                      | Default  | Description                                                                            |
| ------------ | -------- | :------: | --------------------------- | -------- | -------------------------------------------------------------------------------------- |
| type         | `string` |    -     | `submit`, `reset`, `button` | `button` | Used as the HTML `type` attribute.                                                     |
| `label`      | `string` |    ✅    | -                           | -        | Should describe what the button does, for accessibility purposes (screen-reader users) |
| `onClick`    | `func`   |    ✅    | -                           | -        | What the button will trigger when clicked                                              |
| `icon`       | `node`   |    ✅    | -                           | -        | An `Icon` component                                                                    |
| `isDisabled` | `bool`   |    -     | -                           | `false`  | Tells when the button should present a disabled state                                  |

#### Where to use

Mostly in all places where you just need a "clickable" icon, without the complex
behaviours of the `IconButton`

- Pagination list _example: Go to next page_
