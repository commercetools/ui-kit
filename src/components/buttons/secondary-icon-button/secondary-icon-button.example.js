import React from 'react';
import { InformationIcon, SecondaryIconButton } from 'ui-kit';
import { Suite, Spec } from '../../../../test/percy';

export const routePath = '/secondary-icon-button';

export const component = () => (
  <Suite>
    <Spec label="regular">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
      />
    </Spec>
    <Spec label="disabled">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isDisabled={true}
      />
    </Spec>
  </Suite>
);
