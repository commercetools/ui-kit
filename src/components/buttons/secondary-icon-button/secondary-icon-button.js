import React from 'react';
import PropTypes from 'prop-types';
import withMouseOverState from '../../../hocs/with-mouse-over-state';
import filterDataAttributes from '../../../utils/filter-data-attributes';
import AccessibleButton from '../accessible-button';
import styles from './secondary-icon-button.mod.css';

export const SecondaryIconButton = props => {
  const buttonAttributes = {
    'data-track-component': 'SecondaryIconButton',
    ...filterDataAttributes(props),
  };
  let iconTheme = 'black';
  if (props.isDisabled) iconTheme = 'grey';
  else if (props.isMouseOver) iconTheme = 'green';
  return (
    <div
      onMouseOver={props.handleMouseOver}
      onMouseOut={props.handleMouseOut}
      className={styles.button}
    >
      <AccessibleButton
        type={props.type}
        buttonAttributes={buttonAttributes}
        label={props.label}
        onClick={props.onClick}
        isDisabled={props.isDisabled}
      >
        <div className={styles['icon-container']}>
          {React.cloneElement(props.icon, { theme: iconTheme })}
        </div>
      </AccessibleButton>
    </div>
  );
};

SecondaryIconButton.displayName = 'SecondaryIconButton';
SecondaryIconButton.propTypes = {
  type: PropTypes.oneOf(['submit', 'reset', 'button']),
  label: PropTypes.string.isRequired,
  icon: PropTypes.element.isRequired,
  isDisabled: PropTypes.bool,
  onClick: PropTypes.func.isRequired,

  // HoC
  isMouseOver: PropTypes.bool.isRequired,
  handleMouseOver: PropTypes.func.isRequired,
  handleMouseOut: PropTypes.func.isRequired,
};
SecondaryIconButton.defaultProps = {
  type: 'button',
  isDisabled: false,
};

export default withMouseOverState(SecondaryIconButton);
