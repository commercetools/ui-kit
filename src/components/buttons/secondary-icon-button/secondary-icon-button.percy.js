import React from 'react';
import {
  SecondaryIconButton,
  InformationIcon,
} from '../../../../dist/ui-kit.esm';
import { Suite, Spec, screenshot } from '../../../../test/percy';

screenshot('SecondaryIconButton', () => (
  <Suite>
    <Spec label="regular">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
      />
    </Spec>

    <Spec label="disabled">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isDisabled={true}
      />
    </Spec>

    <Spec label='when type is "button"'>
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        type="button"
      />
    </Spec>

    <Spec label='when type is "submit"'>
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        type="submit"
      />
    </Spec>

    <Spec label='when type is "reset"'>
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        type="reset"
      />
    </Spec>
  </Suite>
));
