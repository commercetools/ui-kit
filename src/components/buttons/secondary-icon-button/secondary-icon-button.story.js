import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { withKnobs, boolean, select, text } from '@storybook/addon-knobs/react';
import withReadme from 'storybook-readme/with-readme';
import Section from '../../../../.storybook/decorators/section';
import * as icons from '../../icons';
import Readme from './README.md';
import SecondaryIconButton from './secondary-icon-button';

const iconNames = Object.keys(icons);

storiesOf('Components|Buttons', module)
  .addDecorator(withKnobs)
  .addDecorator(withReadme(Readme))
  .add('SecondaryIconButton', () => (
    <Section>
      <SecondaryIconButton
        icon={React.createElement(
          icons[select('icon', iconNames, iconNames[0])]
        )}
        onClick={action('onClick')}
        label={text('label', 'Accessibility text')}
        isDisabled={boolean('isDisabled', false)}
      />
    </Section>
  ));
