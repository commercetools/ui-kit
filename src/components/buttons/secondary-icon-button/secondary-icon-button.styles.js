import { css } from '@emotion/core';
import vars from '../../../../materials/custom-properties';

const getDisabledStyle = overwrittenVars => {
  /* By using the css 'disabled' selector directly, we don't need additional logic to check the isDisabled prop */
  return css`
    &:disabled svg * {
      fill: ${overwrittenVars.colorNeutral60};
    }
  `;
};

const getColorStyle = (props, overwrittenVars) => {
  switch (props.color) {
    case 'solid':
      return css`
        & svg * {
          fill: ${overwrittenVars.colorSolid};
        }
        &:hover svg * {
          fill: ${overwrittenVars.colorPrimary};
        }
      `;
    case 'primary':
      return css`
        & svg * {
          fill: ${overwrittenVars.colorPrimary};
        }
        &:hover svg * {
          fill: ${overwrittenVars.colorPrimary25};
        }
      `;
    default:
      return css`
        svg * {
          fill: ${overwrittenVars.colorSolid};
        }
      `;
  }
};

const getBaseStyles = (theme, props) => {
  const overwrittenVars = {
    ...vars,
    ...theme,
  };

  return [
    css`
      display: inline-flex;
      align-items: center;
      justify-content: center;
    `,
    getColorStyle(props, overwrittenVars),
    getDisabledStyle(overwrittenVars),
  ];
};

// eslint-disable-next-line import/prefer-default-export
export { getBaseStyles };
