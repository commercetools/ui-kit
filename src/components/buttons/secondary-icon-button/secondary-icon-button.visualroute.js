import React from 'react';
import { ThemeProvider } from 'emotion-theming';
import { InformationIcon, SecondaryIconButton } from 'ui-kit';
import { Suite, Spec } from '../../../../test/percy';

export const routePath = '/secondary-icon-button';

export const component = ({ themes }) => (
  <Suite>
    <Spec label="regular">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
      />
    </Spec>
    <Spec label="disabled">
      <SecondaryIconButton
        icon={<InformationIcon />}
        label="A label text"
        onClick={() => {}}
        isDisabled={true}
      />
    </Spec>
    <ThemeProvider theme={themes.darkTheme}>
      <Spec label="with custom (dark) theme">
        <SecondaryIconButton
          icon={<InformationIcon />}
          label="A label text"
          onClick={() => {}}
        />
      </Spec>
    </ThemeProvider>
  </Suite>
);
