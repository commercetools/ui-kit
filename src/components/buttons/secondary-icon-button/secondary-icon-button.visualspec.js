import { percySnapshot } from '@percy/puppeteer';

describe('SecondaryIconButton', () => {
  beforeAll(async () => {
    await page.goto(`${HOST}/secondary-icon-button`);
  });

  it('Default', async () => {
    await expect(page).toMatch('A label text');
    await percySnapshot(page, 'SecondaryIconButton');
  });
});
