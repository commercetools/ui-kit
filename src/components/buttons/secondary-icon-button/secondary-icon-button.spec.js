import React from 'react';
import { shallow } from 'enzyme';
import { FilterIcon } from '../../icons';
import { SecondaryIconButton } from './secondary-icon-button';
import AccessibleButton from '../accessible-button';

const createTestProps = custom => ({
  label: 'Accessible button',
  onClick: jest.fn(),
  icon: <FilterIcon />,

  // HoC
  isMouseOver: false,
  handleMouseOver: jest.fn(),
  handleMouseOut: jest.fn(),

  ...custom,
});

describe('rendering', () => {
  let props;
  let wrapper;
  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<SecondaryIconButton {...props} />);
  });

  it('outputs correct tree', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render a <AccessibleButton>', () => {
    expect(wrapper).toRender('AccessibleButton');
  });

  it('should render the icon', () => {
    expect(wrapper).toRender('FilterIcon');
  });

  it('forwards the type', () => {
    expect(wrapper.find(AccessibleButton)).toHaveProp('type', 'button');
  });

  it('renders an icon with black theme', () => {
    expect(wrapper.find('FilterIcon')).toHaveProp('theme', 'black');
  });

  describe('if mouse is over', () => {
    beforeEach(() => {
      props = createTestProps({ isMouseOver: true });
      wrapper = shallow(<SecondaryIconButton {...props} />);
    });
    it('renders an icon with green theme', () => {
      expect(wrapper.find('FilterIcon')).toHaveProp('theme', 'green');
    });
  });

  describe('if disabled', () => {
    beforeEach(() => {
      props = createTestProps({ isDisabled: true });
      wrapper = shallow(<SecondaryIconButton {...props} />);
    });
    it('renders a <AccessibleButton> with disabled prop', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp('isDisabled', true);
    });
    it('renders an icon with grey theme', () => {
      expect(wrapper.find('FilterIcon')).toHaveProp('theme', 'grey');
    });
  });

  describe('with data-* props', () => {
    beforeEach(() => {
      props = createTestProps({
        'data-track-component': 'SecondaryIconButton',
        'data-track-label': 'SecondaryIconButton',
        'data-track-event': 'click',
        'data-test': 'secondary-icon-button',
      });
      wrapper = shallow(<SecondaryIconButton {...props} />);
    });
    it('should apply given `data-track-component` to AccessibleButton', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'buttonAttributes',
        expect.objectContaining({
          'data-track-component': 'SecondaryIconButton',
        })
      );
    });
    it('should apply given `data-track-event` to AccessibleButton', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'buttonAttributes',
        expect.objectContaining({
          'data-track-event': 'click',
        })
      );
    });
    it('should apply given `data-track-label` to AccessibleButton', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'buttonAttributes',
        expect.objectContaining({
          'data-track-label': 'SecondaryIconButton',
        })
      );
    });
    it('should apply given `data-test` to AccessibleButton', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'buttonAttributes',
        expect.objectContaining({
          'data-test': 'secondary-icon-button',
        })
      );
    });
  });
  describe('without data-* props', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<SecondaryIconButton {...props} />);
    });
    it('should apply default `data-track-component` to AccessibleButton', () => {
      expect(wrapper.find('AccessibleButton')).toHaveProp(
        'buttonAttributes',
        expect.objectContaining({
          'data-track-component': 'SecondaryIconButton',
        })
      );
    });
  });
});

describe('callbacks', () => {
  let wrapper;
  let props;
  describe('onClick', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<SecondaryIconButton {...props} />);
      wrapper.find('AccessibleButton').prop('onClick')();
    });

    it('should invoke onClick', () => {
      expect(props.onClick).toHaveBeenCalledTimes(1);
    });
  });
});
