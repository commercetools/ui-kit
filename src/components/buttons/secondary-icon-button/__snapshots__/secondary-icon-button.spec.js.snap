// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering outputs correct tree 1`] = `
<div
  className="button"
  onMouseOut={[MockFunction]}
  onMouseOver={[MockFunction]}
>
  <AccessibleButton
    buttonAttributes={
      Object {
        "data-track-component": "SecondaryIconButton",
      }
    }
    isDisabled={false}
    isToggleButton={false}
    label="Accessible button"
    onClick={[MockFunction]}
    type="button"
  >
    <div
      className="icon-container"
    >
      <FilterIcon
        theme="black"
      />
    </div>
  </AccessibleButton>
</div>
`;
